src/shrinkage.jl

```diff
@@ -0,0 +1,84 @@
+"""
+    CoefByGroup
+
+Coefficients for groups of observations defined by levels of a grouping factor
+
+Fields include:
+- `cnames`: `Vector{String}` of the column (or coefficient) names
+- `levels`: `Vector` of the levels of the grouping factor.  Usually `String`s but not necessarily.
+- `fixed`: `AbstractVector{T}` of the global OLS estimates for the coefficients in `cnames` only.
+- `condmodes`: `AbstractMatrix{T}` of size `length(levels) x length(cnames)` of the conditional means/modes for the random effects
+- `grpest`: similar to `condmodes` but allowing for `Missing` values, giving the within group OLS estimates of the coefficients.
+"""
+struct CoefByGroup{T<:AbstractFloat}
+    cnames::Vector{String}
+    levels::Vector
+    fixed::AbstractVector{T}
+    condmodes::AbstractMatrix{T}
+    grpest::AbstractMatrix{Union{Missing,T}}
+end
+
+"""
+    shrinkage(m)
+
+Return a `NamedTuple{fnames(m), NTuple(k, CoefByGroup)}` of the coefficients by group for the grouping factors
+"""
+function shrinkage(m::LinearMixedModel{T}) where {T}
+    fenms = m.feterm.cnames
+    yvec = view(m.Xymat, :, size(m.Xymat, 2))
+    ranefs = ranef(m)
+    cvec = sizehint!(CoefByGroup[], length(ranefs))
+    for (j, re) in enumerate(m.reterms)
+        cnms = re.cnames
+        cnms ⊆ fenms || throw(ArgumentError("re[$j].cnames is not a subset of m.feterm.cnames"))
+        Xmat = view(m.Xymat, :, [findfirst(==(nm), fenms) for nm in cnms])
```

Comment:
This avoids dealing with the formula interface, but it does mean that we can't get shrinkage estimates from a purely random effects model, such as `y ~ 0 + (1+x|g)`. I know that such models can become problematic fast, but

1. They do represent a way to regularize towards zero instead of towards the grand mean. (Okay, not great and not well understand by most users.)
2. I have seen a fair amount of debate on whether you should drop the corresponding random effect when dropping a fixed effect for model comparison, so people are fitting models like `y ~ 1 + (1 + x | g)`. For me, that's quite similar to a Type-III test unless you have strong reasons to assume that the population-level mean is exactly 0.  I'm not a fan of Type-III test but they remain popular in large swathes of psychology and cognitive science.

As a documentation issue, maybe something like:
```suggestion
        cnms ⊆ fenms || throw(ArgumentError("No corresponding fixed effect for random effects $(setdiff(cnms, fenms)): there is no estimated grand mean to measure shrinkage towards"))
        Xmat = view(m.Xymat, :, [findfirst(==(nm), fenms) for nm in cnms])
```

It tells the user which random slopes don't have corresponding fixed effects and why we don't bother estimating shrinkage there.